### PR TITLE
[SECURITY-3075] `getAggregateSecretPattern` to fail when run inside agent JVM

### PR DESCRIPTION
**Pull request by jglick**
Cleans up after #239, especially relevant now given https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/323.

### CODE REVIEW

**Comment by jglick** [comment]
:thinking: Given https://github.com/jenkinsci/workflow-api-plugin/pull/290 I should probably write a test to make sure that the ISE is not just caught and ignored.

File: src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/SecretPatternsTest.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        DurableTaskStep.USE_WATCHING = true;
+        WorkflowJob p = r.createProject(WorkflowJob.class, "p");
+        p.setDefinition(new CpsFlowDefinition("node('remote') {def msg = 'echo do not look at s3cr3t please'; if (isUnix()) {sh msg} else {bat msg}}", true));
+        agents.createAgent(r, "remote");
+        try {
+            WorkflowRun b = r.waitForCompletion(p.scheduleBuild2(0).waitForStart());
+            r.assertLogContains(BadMasker.class.getName(), b);
+            /* Not currently ensured:
+            r.assertLogNotContains("s3cr3t", b);
+            */
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/SecretPatternsTest.java:65]: At first I thought I might have `getAggregateSecretPattern` log a warning but return `Pattern.compile(".+")` to guarantee that the secret (along with everything else) is masked. But this just seemed confusing, and the stack trace would be shown only in the agent log, not the controller log nor build console. So I settled on just throwing the exception, which results in a warning in the agent log but also a clear error in the build log:

```
Started
[Pipeline] Start of Pipeline
[Pipeline] node
Running on remote in /…/agent-work-dirs/remote/workspace/p
[Pipeline] {
[Pipeline] isUnix
[Pipeline] sh
java.lang.IllegalStateException: Not running on the Jenkins controller JVM
	at jenkins.util.JenkinsJVM.checkJenkinsJVM(JenkinsJVM.java:46)
	at org.jenkinsci.plugins.credentialsbinding.masking.SecretPatterns.getAggregateSecretPattern(SecretPatterns.java:57)
	at org.jenkinsci.plugins.credentialsbinding.masking.SecretPatternsTest$BadMasker.decorate(SecretPatternsTest.java:73)
	at org.jenkinsci.plugins.workflow.log.TaskListenerDecorator.decorateAll(TaskListenerDecorator.java:231)
	at org.jenkinsci.plugins.workflow.log.TaskListenerDecorator$DecoratedTaskListener.getLogger(TaskListenerDecorator.java:269)
	at org.jenkinsci.plugins.workflow.log.TaskListenerDecorator$CloseableTaskListener.getLogger(TaskListenerDecorator.java:306)
	at org.jenkinsci.plugins.workflow.steps.durable_task.DurableTaskStep$Execution$NewlineSafeTaskListener.getLogger(DurableTaskStep.java:458)
	at org.jenkinsci.plugins.workflow.steps.durable_task.DurableTaskStep$HandlerImpl.output(DurableTaskStep.java:723)
	at org.jenkinsci.plugins.durabletask.FileMonitoringTask$Watcher.run(FileMonitoringTask.java:611)
	at jenkins.security.ImpersonatingScheduledExecutorService$1.run(ImpersonatingScheduledExecutorService.java:67)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
+ echo do not look at s3cr3t please
do not look at s3cr3t please
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
Finished: SUCCESS
```

This makes it obvious to both end users and the plugin developer that something is broken in the plugin, so it should lead to the bad masking actually getting fixed.

File: src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/SecretPatternsTest.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        } finally {
+            agents.stop("remote");
+        }
+    }
+
+    public static final class BadMasker extends TaskListenerDecorator {
+        @Override public OutputStream decorate(OutputStream logger) throws IOException, InterruptedException {
+            Pattern pattern = SecretPatterns.getAggregateSecretPattern(Set.of("s3cr3t"));
+            return new SecretPatterns.MaskingOutputStream(logger, () -> pattern, "UTF-8");
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/SecretPatternsTest.java:74]: As opposed to 
```suggestion
            return new SecretPatterns.MaskingOutputStream(logger, () -> SecretPatterns.getAggregateSecretPattern(Set.of("s3cr3t")), "UTF-8");
```

as in the code prior to https://github.com/jenkinsci/azure-keyvault-plugin/commit/f46b7fa846a1564119fa51f5797dc4a81d876088 for example. In this case the exception is thrown from a different spot and handled differently:

```
Started
[Pipeline] Start of Pipeline
[Pipeline] node
Running on remote in /…/agent-work-dirs/remote/workspace/p
[Pipeline] {
[Pipeline] isUnix
[Pipeline] sh
java.lang.IllegalStateException: Not running on the Jenkins controller JVM
	at jenkins.util.JenkinsJVM.checkJenkinsJVM(JenkinsJVM.java:46)
	at org.jenkinsci.plugins.credentialsbinding.masking.SecretPatterns.getAggregateSecretPattern(SecretPatterns.java:57)
	at org.jenkinsci.plugins.credentialsbinding.masking.SecretPatternsTest$BadMasker.lambda$decorate$0(SecretPatternsTest.java:73)
	at org.jenkinsci.plugins.credentialsbinding.masking.SecretPatterns$MaskingOutputStream.eol(SecretPatterns.java:93)
	at hudson.console.LineTransformationOutputStream.eol(LineTransformationOutputStream.java:61)
	at hudson.console.LineTransformationOutputStream.write(LineTransformationOutputStream.java:57)
	at java.base/java.io.PrintStream.write(PrintStream.java:528)
	at java.base/java.io.FilterOutputStream.write(FilterOutputStream.java:87)
	at org.jenkinsci.plugins.workflow.steps.durable_task.DurableTaskStep$Execution$NewlineSafeTaskListener$1.write(DurableTaskStep.java:462)
	at java.base/java.io.FilterOutputStream.write(FilterOutputStream.java:137)
	at org.jenkinsci.plugins.workflow.steps.durable_task.DurableTaskStep$Execution$NewlineSafeTaskListener$1.write(DurableTaskStep.java:466)
	at org.apache.commons.io.IOUtils.copyLarge(IOUtils.java:1310)
	at org.apache.commons.io.IOUtils.copy(IOUtils.java:978)
	at org.apache.commons.io.IOUtils.copyLarge(IOUtils.java:1282)
	at org.apache.commons.io.IOUtils.copy(IOUtils.java:953)
	at org.jenkinsci.plugins.workflow.steps.durable_task.DurableTaskStep$HandlerImpl.output(DurableTaskStep.java:737)
	at org.jenkinsci.plugins.durabletask.FileMonitoringTask$Watcher.run(FileMonitoringTask.java:611)
	at jenkins.security.ImpersonatingScheduledExecutorService$1.run(ImpersonatingScheduledExecutorService.java:67)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

with the agent printing a warning

```
WARNING: giving up on watching …
```

**Review by timja — APPROVED** [review]

**Comment by basil** [comment]
Caused https://github.com/jenkinsci/hashicorp-vault-plugin/issues/311 and [JENKINS-71788](https://issues.jenkins.io/browse/JENKINS-71788).

**Pull request merged**